### ISSUE TITLE
_BypassTLS.loseConnection specifies a broken default reason

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5336]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#5336|
|Type|release blocker: regression|
|Created|2011-10-26 11:53:33Z|
|Branch|https://github.com/twisted/twisted/tree/bypasstls-5336|

_BypassTLS.loseConnection defines the default reason as None, but it ought to be a failure object, presumably. failure.Failure(main.CONNECTION_DONE) like everywhere else.

<details><summary>Searchable metadata</summary>

```
trac-id__5336 5336
type__release_blocker__regression release blocker: regression
reporter__therve therve
priority__normal normal
milestone__Twisted_11_1 Twisted-11.1
branch__branches_bypasstls_5336 branches/bypasstls-5336
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1319630013000000 1319630013000000
changetime__1320238387000000 1320238387000000
version__None None
owner__itamar itamar
cc__exarkun cc__glyph
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

I notice that iocpreactor Connection.loseConnection may have the same bug.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Here's the reported traceback which may be caused by this issue:

```
Unhandled Error
Traceback (most recent call last):
  File ".../twisted/python/log.py", line 84, in callWithLogger
   return callWithContext({"system": lp}, func, *args, **kw)
  File ".../twisted/python/log.py", line 69, in callWithContext
   return context.call({ILogContext: newCtx}, func, *args, **kw)
  File ".../twisted/python/context.py", line 118, in callWithContext
   return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File ".../twisted/python/context.py", line 81, in callWithContext
   return func(*args,**kw)
--- <exception caught here> ---
  File ".../twisted/internet/selectreactor.py", line 150, in _doReadOrWrite
   why = getattr(selectable, method)()
  File ".../twisted/internet/tcp.py", line 199, in doRead
   rval = self.protocol.dataReceived(data)
  File ".../twisted/protocols/tls.py", line 413, in dataReceived
   self._flushReceiveBIO()
  File ".../twisted/protocols/tls.py", line 360, in _flushReceiveBIO
   self._shutdownTLS()
  File ".../twisted/protocols/tls.py", line 427, in _shutdownTLS
   self.transport.loseConnection()
  File ".../twisted/internet/_newtls.py", line 71, in loseConnection
   return self._base.loseConnection(self._connection, reason)
  File ".../twisted/internet/abstract.py", line 360, in loseConnection
   self.connectionLost(_connDone)
  File ".../twisted/internet/tcp.py", line 272, in connectionLost
   self._closeSocket(not reason.check(error.ConnectionAborted))
exceptions.AttributeError: 'NoneType' object has no attribute 'check'
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

I opened the similar [#5341](https://github.com/twisted/twisted/issues/5341)... but even if we fix the exception above by changing _BypassTLS, that still leaves us with a broken loseWriteConnection implementation, so having this as a separate ticket is fine, I think.

It's possible the current half-close support is only semi-broken, and we can release 11.1 after only fixing this.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve removed owner|
|-|-|

Mokay, I tried a little bit but I don't manage to create a test (let along a unittest) for this. It seems it should only appear for half-closed protocol, but I can't see how to get there.

One discovery: `TLSMemoryBIOProtocol` claims to `skip ProtocolWrapper.makeConnection`. but by doing it also skips the `directlyProvides(self, providedBy(transport))` call. Also, shouldn't there be a `directlyProvides(self, providedBy(self.wrappedProtocol))` call in `ProtocolWrapper`?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

And yet, somehow, exarkun demonstrated yesterday on IRC that a TLSMemBIOProtocol instances provided IHalfCloseableDescriptor (I think).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

It's worth nothing that TLS connections have two different states: in connectionMade, before any data has been received, you're still in pre-TLS-handshake stage, and so various operations will use different internal code paths as opposed to after dataReceived.

That being said, I tried it both ways and had no luck reproducing either. Maybe we can get glyph to do it by simplifying his code base?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

Half-closeable:

```
>>> from twisted.internet import reactor, protocol
>>> from twisted.internet import reactor, protocol, endpoints
>>> e = endpoints.clientFromString(reactor, 'ssl:host=google.com:port=443')
>>> f = protocol.Factory()
>>> f.protocol = protocol.Protocol
>>> d = e.connect(f)
>>> t = d.result.transport
>>> from twisted.internet import interfaces
>>> interfaces.IHalfCloseableDescriptor.providedBy(t)
True
>>> t.__class__
<class twisted.protocols.tls.TLSMemoryBIOProtocol at 0x93956ec>
>>> 
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst commented|
|-|-|

Since reproducing is impossible, I suggest a different testing approach. I propose rewriting _BypassTLS to use `functools.partial` to implement the 3 _base methods, and then assert that it's doing that by inspecting the results.

Oh, except that's 2.5 only and we're still supporting 2.4 in this release. So I guess instead of `functools.partial` use some other function creator that is testable. The main point being that this way we don't have to manually recreate the signature, and testing with a fake _base is feasible.

Sound OK?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

Wouldn't it be simpler to just pass in a fake for the `base` object?  Say, a `StringTransport`?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst commented|
|-|-|

I'd rather assert "the call is passed through transparently" rather than "we used this specific default argument here".

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

I like assertions like "the desired behavior was achieved".  But go ahead and do what you think is best.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst commented|
|-|-|

(In [33027]) Branching to 'bypasstls-5336'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_14"></a>|@itamarst commented|
|-|-|

Ready for review.

Test run:
http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/bypasstls-5336

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph set owner to @itamarst|
|-|-|

Looks good to me.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_16"></a>|@itamarst set status to `closed`|
|-|-|

(In [33048]) Merge bypasstls-5336.

Author: itamar
Review: glyph
Fixes: [#5336](https://github.com/twisted/twisted/issues/5336)

Preserve default arguments in FileDescriptor.loseConnection when wrapping for TLS.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_17"></a>|@itamarst commented|
|-|-|

Glyph, could you retest your server now that this fix is in?
